madsim-aws-sdk-s3/src/server/service.rs

```diff
@@ -0,0 +1,624 @@
+    }
+
+    pub async fn put_object(
+        &self,
+        bucket: String,
+        key: String,
+        object: crate::types::ByteStream,
+    ) -> Result<PutObjectOutput> {
+        self.timeout().await?;
+        self.inner.lock().put_object(bucket, key, object).await
+    }
+
+    pub async fn delete_object(&self, bucket: String, key: String) -> Result<DeleteObjectOutput> {
+        self.timeout().await?;
+        self.inner.lock().delete_object(bucket, key)
+    }
+
+    pub async fn delete_objects(
+        &self,
+        bucket: String,
+        delete: crate::model::Delete,
+    ) -> Result<DeleteObjectsOutput> {
+        self.timeout().await?;
+        self.inner.lock().delete_objects(bucket, delete)
+    }
+
+    pub async fn head_object(&self, bucket: String, key: String) -> Result<HeadObjectOutput> {
+        self.timeout().await?;
+        self.inner.lock().head_object(bucket, key)
+    }
+
+    pub async fn list_objects_v2(
+        &self,
+        bucket: String,
+        prefix: Option<String>,
+        continuation_token: Option<String>,
+    ) -> Result<ListObjectsV2Output> {
+        self.timeout().await?;
+        self.inner
+            .lock()
+            .list_objects_v2(bucket, prefix, continuation_token)
+    }
+
+    pub async fn get_bucket_lifecycle_configuration(
+        &self,
+        bucket: String,
+        expected_bucket_owner: Option<String>,
+    ) -> Result<GetBucketLifecycleConfigurationOutput> {
+        self.timeout().await?;
+        self.inner
+            .lock()
+            .get_bucket_lifecycle_configuration(bucket, expected_bucket_owner)
+    }
+
+    pub async fn put_bucket_lifecycle_configuration(
+        &self,
+        bucket: String,
+        lifecycle_configuration: Option<BucketLifecycleConfiguration>,
+        expected_bucket_owner: Option<String>,
+    ) -> Result<PutBucketLifecycleConfigurationOutput> {
+        self.timeout().await?;
+        self.inner.lock().put_bucket_lifecycle_configuration(
+            bucket,
+            lifecycle_configuration.unwrap_or(BucketLifecycleConfiguration {
+                rules: Some(Vec::new()),
+            }),
+            expected_bucket_owner,
+        )
+    }
+
+    async fn timeout(&self) -> Result<()> {
+        if thread_rng().gen_bool(self.timeout_rate as f64) {
+            let t = thread_rng().gen_range(Duration::from_secs(5)..Duration::from_secs(15));
+            madsim::time::sleep(t).await;
+            tracing::warn!(?t, "etcdserver: request timed out");
+            return Err(Error::GRpcStatus(tonic::Status::new(
+                tonic::Code::Unavailable,
+                "etcdserver: request timed out",
+            )));
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Default)]
+struct ServiceInner {
+    /// (bucket, key) -> Object
+    storage: BTreeMap<String, BTreeMap<String, InnerObject>>,
+
+    /// (bucket) -> LifecycleRules
+    lifecycle: BTreeMap<String, Vec<LifecycleRule>>,
```

Comment:
It seems that lifecycle is only stored but not used. Will this cause any potential problems for Risingwave?